workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java

```diff
@@ -168,6 +169,64 @@ public WorkReport execute(WorkFlowRequestDTO workFlowRequestDTO) {
 		return new DefaultWorkReport(WorkStatus.IN_PROGRESS, workContext);
 	}
 
+	@Override
+	public WorkReport restart(UUID workFlowExecutionId) {
+		User user = userService.getUserEntityByUsername(SecurityUtils.getUsername());
+		WorkFlowExecution workFlowExecution = workFlowRepository.findById(workFlowExecutionId).orElseThrow(() -> {
+			throw new ResourceNotFoundException(ResourceType.WORKFLOW_EXECUTION, workFlowExecutionId);
+		});
+		WorkFlowDefinition workFlowDefinition = Optional.ofNullable(workFlowExecution.getWorkFlowDefinition())
+				.orElseThrow(() -> {
+					throw new ResourceNotFoundException(ResourceType.WORKFLOW_DEFINITION, workFlowExecution.getId());
+				});
+		WorkFlowDefinitionResponseDTO workFlowDefinitionResponseDTO = workFlowDefinitionService
+				.getWorkFlowDefinitionByName(workFlowDefinition.getName());
+		// check if it is not an inner workflow
+		if (workFlowExecution.getMainWorkFlowExecution() != null) {
+			throw new IllegalWorkFlowStateException(
+					String.format("workflow id: %s from workflow name: %s is an inner workflow!",
+							workFlowExecution.getId(), workFlowDefinition.getName()));
+		}
+		if (workFlowExecution.getWorkFlowExecutionContext() == null) {
+			throw new IllegalWorkFlowStateException(String.format(
+					"workflow id: %s from workflow name: %s has not Workflow Execution Context saved in the database, cannot restart it!",
+					workFlowExecution.getId(), workFlowDefinition.getName()));
+		}
+
+		WorkFlowExecution restartedWorkFlowExecution = saveWorkFlow(workFlowExecution.getProjectId(), user.getId(),
+				workFlowDefinition, WorkStatus.IN_PROGRESS, null, workFlowExecution.getArguments(), workFlowExecution);
+		WorkContext context = rebuildRestartedWorkContext(
```

Comment:
so at this point, the workflow is saved without any workflow-execution-context?
Is there a need to update the workflow with the context or is it done somewhere by the workflow aspect?

Reply:
> so at this point, the workflow is saved without any workflow-execution-context?

Yep, same as `execute`

The context is saved into the workflow by the aspect when they call `workFlowService.updateWorkFlow()` but this is a good point, the context shall now be saved before the execution, as soon as the workflow is saved in the db so, no matter what happens next, the context is there and it can be restarted correctly